todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TaskViewModel.java

```diff
@@ -0,0 +1,146 @@
+import com.example.android.architecture.blueprints.todoapp.data.Task;
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource;
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository;
+
+
+/**
+ * Abstract class for View Models that expose a single {@link Task}.
+ */
+public abstract class TaskViewModel extends BaseObservable
+        implements TasksDataSource.GetTaskCallback, SnackBarChangedCallback.SnackBarViewModel {
+
+    public final ObservableField<String> snackBarText = new ObservableField<>();
+
+    private final TasksRepository mTasksRepository;
+
+    private final Context mContext;
+
+    @Nullable
+    private Task mTask;
+
+    private boolean mIsDataLoading;
+
+    public TaskViewModel(Context context, TasksRepository tasksRepository) {
+        mContext = context;
+        mTasksRepository = tasksRepository;
+    }
+
+    public void start(String taskId) {
+        if (taskId != null) {
+            mIsDataLoading = true;
+            mTasksRepository.getTask(taskId, this);
+        }
+    }
+
+    @Bindable
+    public String getTitle() {
+        if (mTask == null) {
+            return mContext.getString(R.string.no_data);
+        }
+        return mTask.getTitle();
+    }
+
+    @Bindable
+    public String getDescription() {
+        if (mTask == null) {
+            return mContext.getString(R.string.no_data_description);
+        }
+        return mTask.getDescription();
+    }
+
+    @Bindable
+    public boolean getCompleted() {
+        return mTask.isCompleted();
+    }
+
+    public void setCompleted(boolean completed) {
+        completeChanged(completed);
+    }
+
+    @Bindable
+    public boolean isDataAvailable() {
+        return mTask != null;
+    }
+
+    @Bindable
+    public boolean isDataLoading() {
+        return mIsDataLoading;
+    }
+
+    @Bindable
+    public String getTitleForList() {
+        if (mTask == null) {
+            return "No data";
+        }
+        return mTask.getTitleForList();
+    }
+
+    @Override
+    public void onTaskLoaded(Task task) {
+        mTask = task;
+        mIsDataLoading = false;
+        notifyChange();
+    }
+
+    @Override
+    public void onDataNotAvailable() {
+        mTask = null;
+        mIsDataLoading = false;
+    }
+
+    public void deleteTask() {
+        mTasksRepository.deleteTask(mTask.getId());
+    }
+
+    public void onRefresh() {
+        if (mTask != null) {
+            start(mTask.getId());
+        }
+    }
+
+    @Override
+    public String getSnackBarText() {
+        return snackBarText.get();
+    }
+
+    @Nullable
+    protected String getTaskId() {
+        return mTask.getId();
+    }
+
+    private void completeChanged(boolean isChecked) {
+        mTask.setCompleted(isChecked);
+        if (isChecked) {
+            mTasksRepository.completeTask(mTask);
+            snackBarText.set(mContext.getResources().getString(R.string.task_marked_complete));
```

Comment:
As far as I can tell, this is what we need the `Context` for in this class. 

I would change the `snackBarText` form an `ObservableField<String>` to an `ObservableInteger`. In the case where we would need more complex resource retrieval (string plurals etc), we can use an `ObservableField<Pair<Integer, Integer>>` or something to emit the value to the view. The view would then be responsible for the actual retrieval of the string resource.